### PR TITLE
feat: Compatibility modification of megatron-fsdp

### PR DESCRIPTION
**Pull request by shjwudp**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Starting with M-Core v0.14, custom-fsdp was changed to megatron-fsdp, which introduced some API changes. This MR addresses these changes. The most significant change is checkpoint format change from M-Core `torch_dist` to `fsdp_dtensor`.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Review by BoxiangW — DISMISSED** [review]
LGTM, thanks!

File: scripts/performance/llm/pretrain_llama3_8b.py [hunk 1]
```diff
@@ -84,6 +84,49 @@ def override_recipe_configs(
     recipe = set_exp_logging_configs(
         recipe, "pre_train", "llm", "llama3", args.tensorboard, args.wandb, args.wandb_prj_name, args.wandb_job_name
     )
+    # for saving checkpoints
+    ckpt_path = "/lustre/fsw/coreai_devtech_all/jianbinc/playground/nemo_nvfsdp_update/NeMo/checkpoints"
```
Comment (BoxiangW) [inline scripts/performance/llm/pretrain_llama3_8b.py]: We should remove this path

File: nemo/lightning/pytorch/strategies/megatron_strategy.py [hunk 2]
```diff
@@ -365,9 +365,9 @@ def __init__(
 
         self._fsdp = None
 
-        if fsdp is None and self.ddp_config and self.ddp_config.use_custom_fsdp:
```
Comment (gautham-kollu) [inline nemo/lightning/pytorch/strategies/megatron_strategy.py:368]: Customers who built training scripts/tutorials using this would break. We have should comms around it.

Can we should release-note documentation for this ?

File: scripts/performance/llm/pretrain_llama3_8b.py (hunk 1 above)
Comment (shjwudp) [reply scripts/performance/llm/pretrain_llama3_8b.py]: Yes, thanks for the review!

File: nemo/lightning/pytorch/strategies/megatron_strategy.py (hunk 2 above)
Comment (shjwudp) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:368]: Hi, thanks for point it, I've added `use_custom_fsdp` as a fallback so that users using older M-Core won't experience breaks.

Can I ask where should I add a release note?

Comment (gautham-kollu) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:368]: Creating an issue to track this to be done by 25.11.

**Review by gautham-kollu — APPROVED** [review]
